File: apps/ios/Packages/Compat/Sources/Compat/ServerVersion.swift
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
import Foundation
import Model

public struct ServerVersion: Sendable, Comparable {
    public let major: Int
    public let minor: Int
    public let patch: Int
    public let suffix: String

    public init?(_ text: String?) {
        guard let text,
            let match = text.wholeMatch(
                of: /^v?(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)([-+][0-9A-Za-z.+-]+)?$/),
            let major = Int(match.1), let minor = Int(match.2), let patch = Int(match.3)
        else { return nil }
        self.major = major
        self.minor = minor
        self.patch = patch
        self.suffix = match.4.map(String.init) ?? ""
    }

    public static func < (lhs: Self, rhs: Self) -> Bool {
        (lhs.major, lhs.minor, lhs.patch) < (rhs.major, rhs.minor, rhs.patch)
    }

    public var compatibility: Compatibility {
        let numbers = (major, minor, patch)
        if numbers < (3, 4, 1) { return .tooOld }
        if numbers > (3, 4, 1) || !suffix.isEmpty { return .untested }
        return .tested
    }
}

File: apps/android/core/compat/src/main/kotlin/app/gauja/core/compat/ServerVersion.kt
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
package app.gauja.core.compat

import app.gauja.core.model.Compatibility

// The supported baseline is the recorded release in api/README.md; future versions are not tested
// claims.
data class ServerVersion(val major: Int, val minor: Int, val patch: Int, val suffix: String) :
    Comparable<ServerVersion> {
    override fun compareTo(other: ServerVersion): Int =
        compareValuesBy(this, other, { it.major }, { it.minor }, { it.patch })

    fun compatibility(): Compatibility {
        val baseline = ServerVersion(3, 4, 1, "")
        return when {
            this < baseline -> Compatibility.TOO_OLD
            this > baseline || suffix.isNotEmpty() -> Compatibility.UNTESTED
            else -> Compatibility.TESTED
        }
    }

    companion object {
        fun parse(raw: String?): ServerVersion? {
            val match =
                Regex(
                        "^v?(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)([-+][0-9A-Za-z.+-]+)?$"
                    )
                    .matchEntire(raw.orEmpty()) ?: return null
            return ServerVersion(
                match.groupValues[1].toIntOrNull() ?: return null,
                match.groupValues[2].toIntOrNull() ?: return null,
                match.groupValues[3].toIntOrNull() ?: return null,
                match.groupValues[4],
            )
        }
    }
}
